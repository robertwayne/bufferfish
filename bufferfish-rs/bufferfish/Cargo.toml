[package]
name = "bufferfish"
description = "Schemaless binary protocol for game networking."
authors = ["Rob Wagner <rob@sombia.com>"]
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
bufferfish-internal = { path = "../bufferfish-internal" }
bufferfish-derive = { path = "../bufferfish-derive", optional = true }

[features]
default = []
pretty-print = ["bufferfish-internal/unicode-width"]
derive = ["bufferfish-derive"]
