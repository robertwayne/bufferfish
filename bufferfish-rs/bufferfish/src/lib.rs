use std::{
    fs::{File, OpenOptions},
    io::{BufRead, BufReader, Write},
};

#[cfg(feature = "derive")]
pub use bufferfish_derive::*;
pub use bufferfish_internal::*;

fn build_typescript_definitions(lines: &[String]) -> Result<String, std::io::Error> {
    let mut current_value = 0;
    let mut variant_lines = Vec::new();
    let mut inside_enum = false;

    for line in lines {
        let line = line.trim().replace(',', "").to_string();

        if line.starts_with('}') {
            break;
        }

        if line.starts_with("pub enum PacketId {") {
            inside_enum = true;
            continue;
        }

        if !inside_enum {
            continue;
        }

        let parts: Vec<&str> = line.split('=').map(|s| s.trim()).collect();
        let explicit_value = parts.get(1);

        let value = match explicit_value {
            Some(value) => value.to_string(),
            None => current_value.to_string(),
        };

        variant_lines.push(format!("\t{} = {}", parts[0], value));
        current_value += 1;
    }

    let mut lines = variant_lines.join(",\n");
    lines.push(','); // Trailing comma

    let mut contents = String::new();
    contents.push_str("/* AUTOGENERATED BUFFERFISH FILE, DO NOT EDIT */\n\n");
    contents.push_str("/* Make sure your bundler is configured to inline TypeScript enums in order to avoid bloated codegen from the default TypeScript enum behaviour. */\n");
    contents.push_str("export enum PacketId {\n");
    contents.push_str(&lines);
    contents.push_str("\n}\n");

    Ok(contents)
}

fn write_typescript_file(dest: &str, contents: &str) -> Result<(), std::io::Error> {
    let mut file = OpenOptions::new()
        .write(true)
        .truncate(true)
        .create(true)
        .open(dest)?;

    file.write_all(contents.as_bytes())?;

    Ok(())
}

fn read_input_file(src: &str) -> Result<Vec<String>, std::io::Error> {
    let input_file = File::open(src).map_err(|e| {
        eprintln!("Failed to open input file: {}", e);
        e
    })?;
    let reader = BufReader::new(input_file);

    let mut contents = Vec::new();
    for line in reader.lines() {
        contents.push(line?);
    }

    Ok(contents)
}

pub fn transpile(src: &str, dest: &str) -> Result<(), std::io::Error> {
    let lines = read_input_file(src)?;
    let contents = build_typescript_definitions(&lines)?;

    write_typescript_file(dest, &contents)?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_transpilation() {
        let input = r#"
            pub enum PacketId {
                Foo = 0,
                Bar,
                Baz = 255,
            }
        "#;

        let expected = "/* AUTOGENERATED BUFFERFISH FILE, DO NOT EDIT */\n\n/* Make sure your bundler is configured to inline TypeScript enums in order to avoid bloated codegen from the default TypeScript enum behaviour. */\nexport enum PacketId {\n\tFoo = 0,\n\tBar = 1,\n\tBaz = 255,\n}\n";

        let lines = input
            .lines()
            .map(|s| s.to_string())
            .collect::<Vec<String>>();

        let result = build_typescript_definitions(&lines).unwrap();

        assert_eq!(result, expected);
    }
}
