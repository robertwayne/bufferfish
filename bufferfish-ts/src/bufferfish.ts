export class Bufferfish {
    private inner: Uint8Array
    private pos: number
    private reading: boolean
    private capacity: number

    constructor(buf: ArrayBuffer = new ArrayBuffer(0)) {
        this.inner = new Uint8Array(buf)
        this.pos = 0
        this.reading = false
        this.capacity = 1024
    }

    /**
     * Writes a byte array to the internal buffer. Returns the numbers of bytes
     * written to the buffer.
     *
     * This should only be called by the library.
     */
    private write(buf: Uint8Array): number {
        if (
            this.capacity > 0 &&
            (buf.length > this.capacity ||
                this.inner.length + buf.length > this.capacity)
        ) {
            throw new Error("Bufferfish is full")
        }

        this.reading = false

        const tmp = new Uint8Array(this.inner.length + buf.length)
        tmp.set(this.inner, 0)
        tmp.set(buf, this.inner.length)
        this.inner = tmp

        const bytesWritten = buf.length
        this.pos += bytesWritten

        return bytesWritten
    }

    /**
     * Returns the (immutable) internal Uint8Array.
     */
    public view = (): Uint8Array => {
        return this.inner.slice()
    }

    /**
     * Resets the buffer cursor to the start postion when reading after a write.
     *
     * This should only be called by the library.
     */
    private startReading(): void {
        if (this.reading) {
            return
        }

        this.pos = 0
        this.reading = true
    }

    /**
     * Sets the max capacity (in bytes) for the internal buffer.
     * A value of 0 will allow the buffer to grow indefinitely.
     */
    public setMaxCapacity(capacity: number): void {
        this.capacity = capacity
    }

    /**
     * Returns the next byte in the buffer without advancing the cursor. Returns
     * undefined if the cursor is at the end of the buffer.
     */
    public peek = (): number | null => {
        this.startReading()

        if (this.pos >= this.inner.length) {
            return null
        }

        return this.inner.slice(this.pos, this.pos + 1)[0] ?? null
    }

    /**
     * Returns the next n bytes in the buffer without advancing the cursor.
     * Returns undefined if the cursor is at the end of the buffer.
     */
    public peekN = (n: number): Uint8Array | null => {
        this.startReading()

        if (this.pos + n > this.inner.length) {
            return null
        }

        return this.inner.slice(this.pos, this.pos + n)
    }

    /**
     * Appends another Bufferfish, Uint8Array, ArrayBuffer, or Array<number> to
     * the buffer. This modifies the Bufferfish in-place.
     */
    public push = (
        arr: Bufferfish | Uint8Array | ArrayBuffer | Array<number>,
    ): void => {
        if (arr instanceof Bufferfish) {
            this.write(arr.view())
        } else if (arr instanceof Uint8Array) {
            this.write(arr)
        } else if (arr instanceof ArrayBuffer) {
            this.write(new Uint8Array(arr))
        } else if (arr instanceof Array) {
            this.write(new Uint8Array(arr))
        } else {
            throw new Error("Invalid type")
        }
    }

    /**
     * Writes a single u8 to the buffer as one byte.
     */
    public writeUint8 = (value: number): void => {
        if (value > 255 || value < 0) {
            throw new Error("Value is out of range")
        }

        const slice: Uint8Array = new Uint8Array(1)
        const view = new DataView(slice.buffer)
        view.setUint8(0, value)

        this.write(slice)
    }

    /**
     * Writes a u16 to the buffer as two bytes.
     */
    public writeUint16 = (value: number): void => {
        if (value > 65535 || value < 0) {
            throw new Error("Value is out of range")
        }

        const slice: Uint8Array = new Uint8Array(2)
        const view = new DataView(slice.buffer)
        view.setUint16(0, value)

        this.write(slice)
    }

    /**
     * Writes a u32 to the buffer as four bytes.
     */
    public writeUint32 = (value: number): void => {
        if (value > 4294967295 || value < 0) {
            throw new Error("Value is out of range")
        }

        const slice: Uint8Array = new Uint8Array(4)
        const view = new DataView(slice.buffer)
        view.setUint32(0, value)

        this.write(slice)
    }

    /**
     * Writes an i8 to the buffer as one byte.
     */
    public writeInt8 = (value: number): void => {
        if (value > 127 || value < -128) {
            throw new Error("Value is out of range")
        }

        const slice: Uint8Array = new Uint8Array(1)
        const view = new DataView(slice.buffer)
        view.setInt8(0, value)

        this.write(slice)
    }

    /**
     * Writes an i16 to the buffer as two bytes.
     */
    public writeInt16 = (value: number): void => {
        if (value > 32767 || value < -32768) {
            throw new Error("Value is out of range")
        }

        const slice: Uint8Array = new Uint8Array(2)
        const view = new DataView(slice.buffer)
        view.setInt16(0, value)

        this.write(slice)
    }

    /**
     * Writes an i32 to the buffer as four bytes.
     */
    public writeInt32 = (value: number): void => {
        if (value > 2147483647 || value < -2147483648) {
            throw new Error("Value is out of range")
        }

        const slice: Uint8Array = new Uint8Array(4)
        const view = new DataView(slice.buffer)
        view.setInt32(0, value)

        this.write(slice)
    }

    /**
     * Writes a bool to the buffer as a single byte.
     */
    public writeBool = (value: boolean): void => {
        this.writeUint8(value ? 1 : 0)
    }

    /**
     * Writes a series of bools to the buffer as a single byte. This allows up
     * to 4 bools to be represented as a single byte. The first 4 bits are used
     * as a mask to determine which of the last 4 bits are set.
     */
    public writePackedBools = (values: Array<boolean>): void => {
        if (values.length > 4) {
            throw new Error(
                "Each packed bool can only represent 4 or fewer values",
            )
        }

        let packed_value = 0x00
        for (const value of values) {
            packed_value <<= 1
            if (value) {
                packed_value |= 1
            }
        }

        this.writeUint8(packed_value)
    }

    /**
     * Writes a unicode string literal to the buffer. It will be prefixed with
     * its length in bytes as a u16 (two bytes), and each character will be 1 to
     * 4-bytes, whichever is the smallest it can fit into.
     */
    public writeString = (value: string): void => {
        const slice: Uint8Array = new TextEncoder().encode(value)

        this.writeUint16(slice.length)
        this.write(slice)
    }

    /**
     * Writes a unicode string literal to the buffer without a length prefix.
     * Each character will be 1 to 4-bytes, whichever is the smallest it can fit
     * into.
     */
    public writeSizedString = (value: string): void => {
        const slice: Uint8Array = new TextEncoder().encode(value)

        this.write(slice)
    }

    /**
     * Writes an array of raw bytes to the buffer. Useful for serializing ///
       distinct structs into byte arrays and appending them to a buffer later.
     */
    public writeRawBytes = (value: Uint8Array): void => {
        this.write(value)
    }

    /**
     * Reads a u8 from the buffer.
     */
    public readUint8 = (): number => {
        this.startReading()

        const buf = new Uint8Array(1)
        buf.set(this.inner.subarray(this.pos, this.pos + 1))
        this.pos += 1

        return new DataView(buf.buffer).getUint8(0)
    }

    /**
     * Reads a u16 from the buffer.
     */
    public readUint16 = (): number => {
        this.startReading()

        const buf = new Uint8Array(2)
        buf.set(this.inner.subarray(this.pos, this.pos + 2))
        this.pos += 2

        return new DataView(buf.buffer).getUint16(0)
    }

    /**
     * Reads a u32 from the buffer.
     */
    public readUint32 = (): number => {
        this.startReading()

        const buf = new Uint8Array(4)
        buf.set(this.inner.subarray(this.pos, this.pos + 4))
        this.pos += 4

        return new DataView(buf.buffer).getUint32(0)
    }

    /**
     * Reads an i8 from the buffer.
     */
    public readInt8 = (): number => {
        this.startReading()

        const buf = new Uint8Array(1)
        buf.set(this.inner.subarray(this.pos, this.pos + 1))
        this.pos += 1

        return new DataView(buf.buffer).getInt8(0)
    }

    /**
     * Reads an i16 from the buffer.
     */
    public readInt16 = (): number => {
        this.startReading()

        const buf = new Uint8Array(2)
        buf.set(this.inner.subarray(this.pos, this.pos + 2))
        this.pos += 2

        return new DataView(buf.buffer).getInt16(0)
    }

    /**
     * Reads an i32 from the buffer.
     */
    public readInt32 = (): number => {
        this.startReading()

        const buf = new Uint8Array(4)
        buf.set(this.inner.subarray(this.pos, this.pos + 4))
        this.pos += 4

        return new DataView(buf.buffer).getInt32(0)
    }

    /**
     * Reads a bool from the buffer.
     */
    public readBool = (): boolean => {
        this.startReading()

        const buf = new Uint8Array(1)
        buf.set(this.inner.subarray(this.pos, this.pos + 1))
        this.pos += 1

        return buf[0] === 1
    }

    /**
     *
     */
    public readPackedBools = (): Array<boolean> => {
        return []
    }

    /**
     * Reads a variable length string from the buffer.
     */
    public readString = (): string => {
        this.startReading()

        const len = this.readUint16()
        const slice = this.inner.subarray(this.pos, this.pos + len)
        const str = new TextDecoder("utf-8").decode(slice)
        this.pos += len

        return str
    }

    /**
     * Reads a sized string from the buffer. You must pass the length of the
     * string in bytes.
     */
    public readSizedString = (size: number): string => {
        this.startReading()

        const slice = this.inner.subarray(this.pos, this.pos + size)
        const str = new TextDecoder("utf-8").decode(slice)
        this.pos += size

        return str
    }

    /**
     * Reads a sized string from the buffer. This will read from the buffers
     * current position until the end of the buffer, so this function should not
     * be used unless you know that the string is the last value in the buffer.
     * This removes the overhead of a length prefix; it is recommended to plan
     * your packets out such that they end with a sized string where possible.
     */
    public readStringRemaining = (): string => {
        this.startReading()

        const slice = this.inner.subarray(this.pos, this.inner.length)
        const str = new TextDecoder("utf-8").decode(slice)
        this.pos = this.inner.length

        return str
    }

    // public serialize = (obj: object) => {}

    // public serializeNumber = (number: number) => {}

    // public serializeString = (string: string) => {}

    // public serializeBoolean = (bool: boolean) => {}
}
