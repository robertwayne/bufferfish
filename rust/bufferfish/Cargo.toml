[package]
name = "bufferfish"
authors = ["Rob Wagner <rob@sombia.com>"]
license = "MIT OR Apache-2.0"
description = "Binary message utility library for Rust and TypeScript interop."
repository = "https://github.com/robertwayne/bufferfish"
categories = ["network-programming"]
keywords = ["binary-protocol", "typescript"]
readme = "README.md"
version = "0.1.0"
edition = "2024"

[features]
default = []
pretty-print = ["bufferfish-core/unicode-width"]
derive = ["bufferfish-derive", "syn"]

[dependencies]
bufferfish-core = { version = "0.1", path = "../bufferfish-core" }
bufferfish-derive = { version = "0.1", path = "../bufferfish-derive", optional = true }
syn = { version = "2", features = ["full"], optional = true }

[dev-dependencies]
trybuild = "1"
