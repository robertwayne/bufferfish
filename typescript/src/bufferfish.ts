const OVERFLOW_ERR = "attempted to read past the end of the Bufferfish"

/**
 * A wrapper around Uint8Array that provides a simple API for reading and
 * writing binary data. This is meant to be used with its companion library in
 * Rust to provide a consistent encoding and decoding.
 *
 * In general, you should prefer to use the generated read* and write* functions
 * that are generated by the `bufferfish` compiler, rather than manually using
 * methods on this class directly.
 */
export class Bufferfish {
    private inner: Uint8Array
    private position: number
    private reading: boolean
    private capacity: number

    private textDecoder: TextDecoder | undefined
    private textEncoder: TextEncoder | undefined

    constructor(bf: ArrayBuffer = new ArrayBuffer(0)) {
        this.inner = new Uint8Array(bf)
        this.position = 0
        this.reading = false
        this.capacity = 1024

        this.textDecoder = undefined
        this.textEncoder = undefined
    }

    /**
     * Writes a byte array to the internal buffer. Returns the numbers of bytes
     * written to the buffer or an error if the capacity is exceeded.
     *
     * This should only be called by the library.
     */
    private write(bf: Uint8Array): number | Error {
        if (
            this.capacity > 0 &&
            (bf.length > this.capacity ||
                this.inner.length + bf.length > this.capacity)
        ) {
            return new Error(
                `Bufferfish capacity exceeded (${this.capacity} bytes)`,
            )
        }

        this.reading = false

        const tmp = new Uint8Array(this.inner.length + bf.length)
        tmp.set(this.inner, 0)
        tmp.set(bf, this.inner.length)
        this.inner = tmp

        const bytesWritten = bf.length
        this.position += bytesWritten

        return bytesWritten
    }

    /**
     * Returns the (immutable) internal Uint8Array.
     */
    public view = (): Uint8Array => {
        return this.inner.slice()
    }

    /**
     * Resets the buffer cursor to the start postion when reading after a write.
     *
     * This should only be called by the library.
     */
    private startReading(): void {
        if (this.reading) {
            return
        }

        this.position = 0
        this.reading = true
    }

    /**
     * Sets the max capacity (in bytes) for the internal buffer.
     * A value of 0 will allow the buffer to grow indefinitely.
     */
    public setMaxCapacity(capacity: number): void {
        this.capacity = capacity
    }

    /**
     * Returns the next byte in the buffer without advancing the cursor.
     *
     * Throws if the cursor is at the end of the buffer.
     */
    public peek = (): number | Error => {
        this.startReading()

        const value = this.inner.slice(this.position, this.position + 1)[0]

        if (this.position >= this.inner.length || value === undefined) {
            return new Error(
                `peek of 1 byte exceeds the max capacity of ${this.capacity} bytes on this Bufferfish`,
            )
        }

        return value
    }

    /**
     * Returns the next n bytes in the buffer without advancing the cursor.
     * Returns undefined if the cursor is at the end of the buffer.
     */
    public peekN = (n: number): Uint8Array | Error => {
        this.startReading()

        const value = this.inner.slice(this.position, this.position + n)

        if (this.position + n > this.inner.length) {
            return new Error(
                `peek of ${n} bytes exceeds the max capacity of ${this.capacity} bytes on this Bufferfish`,
            )
        }

        return value
    }

    /**
     * Appends another `Bufferfish`, `Uint8Array`, `ArrayBuffer`, or `Array<number>` to the buffer. This modifies the `Bufferfish` in-place.
     */
    public push = (
        arr: Bufferfish | Uint8Array | ArrayBuffer | Array<number>,
    ): void | Error => {
        if (arr instanceof Bufferfish) {
            const err = this.write(arr.view())
            if (err instanceof Error) {
                return err
            }
        } else if (arr instanceof Uint8Array) {
            const err = this.write(arr)
            if (err instanceof Error) {
                return err
            }
        } else if (arr instanceof ArrayBuffer) {
            const err = this.write(new Uint8Array(arr))
            if (err instanceof Error) {
                return err
            }
        } else if (arr instanceof Array) {
            const err = this.write(new Uint8Array(arr))
            if (err instanceof Error) {
                return err
            }
        } else {
            return new Error("invalid type")
        }
    }

    /**
     * Writes a single u8 to the buffer as one byte.
     *
     * Returns an error if the value is out of range (0-255).
     */
    public writeUint8 = (value: number): void | Error => {
        if (value > 255 || value < 0) {
            return new Error(`value ${value} must be between 0 and 255`)
        }

        const slice: Uint8Array = new Uint8Array(1)
        const view = new DataView(slice.buffer)
        view.setUint8(0, value)

        const err = this.write(slice)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes a u16 to the buffer as two bytes.
     *
     * Returns an error if the value is out of range (0-65535).
     */
    public writeUint16 = (value: number): void | Error => {
        if (value > 65535 || value < 0) {
            return new Error(`value ${value} must be between 0 and 65535`)
        }

        const slice: Uint8Array = new Uint8Array(2)
        const view = new DataView(slice.buffer)
        view.setUint16(0, value)

        const err = this.write(slice)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes a u32 to the buffer as four bytes.
     *
     * Returns an error if the value is out of range (0-4294967295).
     */
    public writeUint32 = (value: number): void | Error => {
        if (value > 4294967295 || value < 0) {
            return new Error(`value ${value} must be between 0 and 4294967295`)
        }

        const slice: Uint8Array = new Uint8Array(4)
        const view = new DataView(slice.buffer)
        view.setUint32(0, value)

        const err = this.write(slice)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes an i8 to the buffer as one byte.
     *
     * Returns an error if the value is out of range (-128-127).
     */
    public writeInt8 = (value: number): void | Error => {
        if (value > 127 || value < -128) {
            return new Error(`value ${value} must be between -128 and 127`)
        }

        const slice: Uint8Array = new Uint8Array(1)
        const view = new DataView(slice.buffer)
        view.setInt8(0, value)

        const err = this.write(slice)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes an i16 to the buffer as two bytes.
     *
     * Returns an error if the value is out of range (-32768-32767).
     */
    public writeInt16 = (value: number): void | Error => {
        if (value > 32767 || value < -32768) {
            return new Error(`value ${value} must be between -32768 and 32767`)
        }

        const slice: Uint8Array = new Uint8Array(2)
        const view = new DataView(slice.buffer)
        view.setInt16(0, value)

        const err = this.write(slice)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes an i32 to the buffer as four bytes.
     *
     * Returns an error if the value is out of range (-2147483648-2147483647).
     */
    public writeInt32 = (value: number): void | Error => {
        if (value > 2147483647 || value < -2147483648) {
            return new Error(
                `value ${value} must be between -2147483648 and 2147483647`,
            )
        }

        const slice: Uint8Array = new Uint8Array(4)
        const view = new DataView(slice.buffer)
        view.setInt32(0, value)

        const err = this.write(slice)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes a boolean to the buffer as one byte.
     */
    public writeBool = (value: boolean): void | Error => {
        const err = this.writeUint8(value ? 1 : 0)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes a packed array of booleans to the buffer as a single byte.
     * Can pack up to 8 booleans into a single byte.
     */
    public writePackedBools = (values: Array<boolean>): void | Error => {
        if (values.length > 8) {
            return new Error("cannot pack more than 8 bools into a single byte")
        }

        let packedValue = 0x00
        for (const value of values) {
            packedValue <<= 1
            if (value) {
                packedValue |= 1
            }
        }

        packedValue <<= 8 - values.length // Shift remaining bits to the left to fill the byte

        const err = this.writeUint8(packedValue)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Writes a unicode string literal to the buffer. It will be prefixed with
     * its length in bytes as a u16 (two bytes), and each character will be 1 to
     * 4-bytes, whichever is the smallest it can fit into.
     */
    public writeString = (value: string): void | Error => {
        if (!this.textEncoder) this.textEncoder = new TextEncoder()
        const slice: Uint8Array = this.textEncoder.encode(value)

        const err = this.writeUint16(slice.length)
        if (err instanceof Error) {
            return err
        }

        const err2 = this.write(slice)
        if (err2 instanceof Error) {
            return err2
        }
    }

    /**
     * Writes an array of raw bytes to the buffer. Useful for encoding
     * distinct structs into byte arrays and appending them to a buffer later.
     */
    public writeRawBytes = (value: Uint8Array): void | Error => {
        const err = this.write(value)
        if (err instanceof Error) {
            return err
        }
    }

    /**
     * Attempts to read a u8 from the buffer.
     */
    public readUint8(): number | Error {
        this.startReading()

        if (this.position + 1 > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        const value = new DataView(
            this.inner.buffer,
            this.position,
            1,
        ).getUint8(0)
        this.position += 1

        return value
    }

    /**
     * Attempts to read a u16 from the buffer.
     */
    public readUint16(): number | Error {
        this.startReading()

        if (this.position + 2 > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        const value = new DataView(
            this.inner.buffer,
            this.position,
            2,
        ).getUint16(0)
        this.position += 2

        return value
    }

    /**
     * Attempts to read a u32 from the buffer.
     */
    public readUint32(): number | Error {
        this.startReading()

        if (this.position + 4 > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        const value = new DataView(
            this.inner.buffer,
            this.position,
            4,
        ).getUint32(0)
        this.position += 4

        return value
    }

    /**
     * Attempts to read an i8 from the buffer.
     */
    public readInt8(): number | Error {
        this.startReading()

        if (this.position + 1 > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        const value = new DataView(this.inner.buffer, this.position, 1).getInt8(
            0,
        )
        this.position += 1

        return value
    }

    /**
     * Attempts to read an i16 from the buffer.
     */
    public readInt16(): number | Error {
        this.startReading()

        if (this.position + 2 > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        const value = new DataView(
            this.inner.buffer,
            this.position,
            2,
        ).getInt16(0)
        this.position += 2

        return value
    }

    /**
     * Attempts to read an i32 from the buffer.
     */
    public readInt32(): number | Error {
        this.startReading()

        if (this.position + 4 > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        const value = new DataView(
            this.inner.buffer,
            this.position,
            4,
        ).getInt32(0)
        this.position += 4

        return value
    }

    /**
     * Attempts to read a bool from the buffer.
     */
    public readBool(): boolean | Error {
        const valueOrError = this.readUint8()
        if (valueOrError instanceof Error) {
            return valueOrError
        }

        return valueOrError === 1
    }

    /**
     * Attempts to read a packed array of booleans from the buffer.
     * By default, it will read 8 booleans from a single byte. If you
     * packed less than 8, the count parameter can be used to specify
     * how many booleans to read.
     */
    public readPackedBools(count: number = 8): Array<boolean> | Error {
        if (count > 8) {
            return new Error("cannot read more than 8 bools from a single byte")
        }

        const packedValueOrError = this.readUint8()
        if (packedValueOrError instanceof Error) {
            return packedValueOrError
        }

        const packedValue = packedValueOrError as number
        const bools: Array<boolean> = []
        for (let i = 0; i < count; i++) {
            bools.push((packedValue & (1 << (7 - i))) !== 0)
        }

        return bools
    }

    /**
     * Attempts to read a variable length string from the buffer.
     */
    public readString(): string | Error {
        const lengthOrError = this.readUint16()
        if (lengthOrError instanceof Error) {
            return lengthOrError
        }

        const length = lengthOrError as number
        if (this.position + length > this.inner.length) {
            return new Error(OVERFLOW_ERR)
        }

        if (!this.textDecoder) this.textDecoder = new TextDecoder("utf-8")

        const value = this.textDecoder.decode(
            this.inner.subarray(this.position, this.position + length),
        )
        this.position += length

        return value
    }

    /**
     * Attempts to read a variable-length array of elements from the buffer.
     */
    public readArray<T>(readFn: () => T): Array<T> | Error {
        const lengthOrError = this.readUint16()
        if (lengthOrError instanceof Error) {
            return lengthOrError
        }

        const length = lengthOrError as number
        const values: Array<T> = []
        for (let i = 0; i < length; i++) {
            const valueOrError = readFn()
            if (valueOrError instanceof Error) {
                return valueOrError
            }

            values.push(valueOrError as T)
        }

        return values
    }
}
