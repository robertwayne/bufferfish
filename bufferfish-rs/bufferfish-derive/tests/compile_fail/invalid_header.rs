use bufferfish::Serialize;

#[derive(Serialize, Debug)]
#[bufferfish("Hello, world!")]
struct Foo {
    bar: u8,
}

fn main() {}
