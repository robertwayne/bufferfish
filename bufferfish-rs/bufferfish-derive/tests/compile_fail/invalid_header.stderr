error[E0277]: the trait bound `u8: From<&str>` is not satisfied
 --> tests/compile_fail/invalid_header.rs:3:10
  |
3 | #[derive(Serialize, Debug)]
  |          ^^^^^^^^^ the trait `From<&str>` is not implemented for `u8`, which is required by `&str: Into<_>`
  |
  = help: the following other types implement trait `From<T>`:
            <u8 as From<bool>>
            <u8 as From<NonZero<u8>>>
            <u8 as From<Char>>
  = note: required for `&str` to implement `Into<u8>`
  = note: this error originates in the derive macro `Serialize` (in Nightly builds, run with -Z macro-backtrace for more info)
